bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/StateDescriptionServiceImpl.java

```diff
@@ -0,0 +1,109 @@
+    public void addStateDescriptionProvider(StateDescriptionProvider provider) {
+        stateDescriptionProviders.add(provider);
+    }
+
+    public void removeStateDescriptionProvider(StateDescriptionProvider provider) {
+        stateDescriptionProviders.remove(provider);
+    }
+
+    @Override
+    public StateDescription getStateDescription(String itemName, Locale locale) {
+        BigDecimal minimum = null;
+        BigDecimal maximum = null;
+        BigDecimal step = null;
+        String pattern = null;
+        boolean isReadOnly = false;
+        List<StateOption> stateOptions = Collections.emptyList();
+
+        boolean foundNonNullStateDescription = false;
+
+        for (StateDescriptionProvider stateDescriptionProvider : stateDescriptionProviders) {
+            StateDescription currentDescription = stateDescriptionProvider.getStateDescription(itemName, locale);
+            if (currentDescription != null) {
+                foundNonNullStateDescription = true;
+                if (shouldCheckNextDescriptionProvider(minimum, maximum, step, pattern, isReadOnly, stateOptions)) {
+                    if (minimum == null && currentDescription.getMinimum() != null) {
+                        minimum = currentDescription.getMinimum();
+                    }
+                    if (maximum == null && currentDescription.getMaximum() != null) {
+                        maximum = currentDescription.getMaximum();
+                    }
+                    if (step == null && currentDescription.getStep() != null) {
+                        step = currentDescription.getStep();
+                    }
+                    if (pattern == null && currentDescription.getPattern() != null) {
+                        pattern = currentDescription.getPattern();
+                    }
+
+                    // This could potentially override the info from ChannelStateDescription if a
+                    // third StateDescriptionProvider with ranking lower than -1 is added
+                    if (isReadOnly == false) {
+                        isReadOnly = currentDescription.isReadOnly();
+                    }
+
+                    if (stateOptions.isEmpty() && !currentDescription.getOptions().isEmpty()) {
+                        stateOptions = currentDescription.getOptions();
+                    }
+                }
+            }
+        }
+        StateDescription result = foundNonNullStateDescription
+                ? new StateDescription(minimum, maximum, step, pattern, isReadOnly, stateOptions)
+                : null;
+        return result;
+    }
+
+    private static boolean shouldCheckNextDescriptionProvider(BigDecimal minimum, BigDecimal maximum, BigDecimal step,
+            String pattern, boolean isReadOnly, List<StateOption> stateOptions) {
+        return minimum == null || maximum == null || step == null || pattern == null || isReadOnly == false
```

Comment:
Should we document somewhere that if you want to create a "winning" `StateDescriptionProvider` that you have to set not only its priority to the highest value but **in addition* also provide values for **all* fields, i.e. minimum, maximum, step, pattern, isReadOnly + stateOptions?

Reply:
Imho it there should not be such a "winning" ConfigDescriptionProvider: The Channel config description is always to be taken first, so if this one defines all fields from the beginning other would never be able to overwrite. This would result in the pattern from *.items files or *.sitemap files never to be taken into account.

My impression is that this PR should at first just introduce the new StateDescriptionService without changing the merging logic.